managed/services/agents/service_info_broker.go

```diff
@@ -0,0 +1,235 @@
+	var request *agentpb.ServiceInfoRequest
+	switch service.ServiceType {
+	case models.MySQLServiceType:
+		tdp := agent.TemplateDelimiters(service)
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_MYSQL_SERVICE,
+			Dsn:     agent.DSN(service, 2*time.Second, service.DatabaseName, nil),
+			Timeout: durationpb.New(3 * time.Second),
+			TextFiles: &agentpb.TextFiles{
+				Files:              agent.Files(),
+				TemplateLeftDelim:  tdp.Left,
+				TemplateRightDelim: tdp.Right,
+			},
+			TlsSkipVerify: agent.TLSSkipVerify,
+		}
+	case models.PostgreSQLServiceType:
+		tdp := agent.TemplateDelimiters(service)
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_POSTGRESQL_SERVICE,
+			Dsn:     agent.DSN(service, 2*time.Second, service.DatabaseName, nil),
+			Timeout: durationpb.New(3 * time.Second),
+			TextFiles: &agentpb.TextFiles{
+				Files:              agent.Files(),
+				TemplateLeftDelim:  tdp.Left,
+				TemplateRightDelim: tdp.Right,
+			},
+		}
+	case models.MongoDBServiceType:
+		tdp := agent.TemplateDelimiters(service)
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_MONGODB_SERVICE,
+			Dsn:     agent.DSN(service, 2*time.Second, service.DatabaseName, nil),
+			Timeout: durationpb.New(3 * time.Second),
+			TextFiles: &agentpb.TextFiles{
+				Files:              agent.Files(),
+				TemplateLeftDelim:  tdp.Left,
+				TemplateRightDelim: tdp.Right,
+			},
+		}
+	case models.ProxySQLServiceType:
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_PROXYSQL_SERVICE,
+			Dsn:     agent.DSN(service, 2*time.Second, service.DatabaseName, nil),
+			Timeout: durationpb.New(3 * time.Second),
+		}
+	case models.ExternalServiceType:
+		exporterURL, err := agent.ExporterURL(q)
+		if err != nil {
+			return nil, err
+		}
+
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_EXTERNAL_SERVICE,
+			Dsn:     exporterURL,
+			Timeout: durationpb.New(3 * time.Second),
+		}
+	case models.HAProxyServiceType:
+		exporterURL, err := agent.ExporterURL(q)
+		if err != nil {
+			return nil, err
+		}
+
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_HAPROXY_SERVICE,
+			Dsn:     exporterURL,
+			Timeout: durationpb.New(3 * time.Second),
+		}
+	default:
+		return nil, errors.Errorf("unhandled Service type %s", service.ServiceType)
+	}
+	return request, nil
+}
+
+// GetInfoFromService sends a request to pmm-agent to query information from a service.
+func (c *ServiceInfoBroker) GetInfoFromService(ctx context.Context, q *reform.Querier, service *models.Service, agent *models.Agent) error {
+	l := logger.Get(ctx)
+	start := time.Now()
+	defer func() {
+		if dur := time.Since(start); dur > 4*time.Second {
+			l.Warnf("GetInfoFromService took %s.", dur)
+		}
+	}()
+
+	pmmAgentID := pointer.GetString(agent.PMMAgentID)
```

Comment:
btw agent.PMMAgentID might be empty for some agents

Reply:
and it might cause errors

Reply:
How come? If you mean it can be empty when ServiceType equals `ExternalServiceType` or `HAProxyServiceType`, then we have a guard for it.

Reply:
ah, you are right